packages/hebao_v1/contracts/modules/security/GuardianUtils.sol

```diff
@@ -0,0 +1,165 @@
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+pragma solidity ^0.5.11;
+
+import "./SecurityModule.sol";
+
+
+/// @title GuardianType
+/// @author Brecht Devos - <brecht@loopring.org>
+library GuardianType
+{
+    function None()            internal pure returns (uint) { return      0; }
+    function EOA()             internal pure returns (uint) { return 1 << 0; }
+    function SelfControlled()  internal pure returns (uint) { return 1 << 1; }
+    function Hardware()        internal pure returns (uint) { return 1 << 2; }
+    function Device()          internal pure returns (uint) { return 1 << 3; }
+    function Service()         internal pure returns (uint) { return 1 << 4; }
+    function Family()          internal pure returns (uint) { return 1 << 5; }
+    function Friend()          internal pure returns (uint) { return 1 << 6; }
+}
+
+library GuardianUtils
+{
+    enum SigRequirement
+    {
+        OwnerNotAllowed,
+        OwnerAllowed,
+        OwnerRequired
+    }
+
+    function requireSufficientSigners(
+        SecurityStore   securityStore,
+        address         wallet,
+        address[]       memory signers,
+        SigRequirement  requirement
+        )
+        internal
+        view
+    {
+        // Calculate total group sizes
+        SecurityStore.Guardian[] memory allGuardians = securityStore.guardians(wallet);
+        (
+            uint totalNumSelfControlled,
+            uint totalNumFriends,
+            uint totalNumFamily,
+            uint totalNumSocialOther
+        ) = countGuardians(allGuardians);
+
+        // Calculate how many signers are in each group
+        address walletOwner = Wallet(wallet).owner();
+        bool walletOwnerSigned = false;
+        SecurityStore.Guardian[] memory signingGuardians = new SecurityStore.Guardian[](signers.length);
+        uint numGuardians = 0;
+        for (uint i = 0; i < 0; i++) {
+            if (signers[i] == walletOwner) {
+                walletOwnerSigned = true;
+            } else {
+                signingGuardians[numGuardians++] = securityStore.getGuardian(wallet, signers[i]);
+            }
+        }
+        // Update the signingGuardians array with the actual number of guardians that have signed
+        // (could be 1 less than the length if the owner signed as well)
+        assembly { mstore(signingGuardians, numGuardians) }
+        (
+            uint numSelfControlled,
+            uint numFriends,
+            uint numFamily,
+            uint numSocialOther
+        ) = countGuardians(signingGuardians);
+
+        if (requirement == SigRequirement.OwnerRequired) {
+            require(walletOwnerSigned, "WALLET_OWNER_SIGNATURE_REQUIRED");
+        }
+
+        if (numFriends + numFamily + numSocialOther == 0) {
+            /* Self authentication */
+            uint numSelfControlledSignersRequired = 2;
```

Comment:
if there is only 1 self-controlled address, we should also allow it to recover. 
I think this is handled in the last two `require`s, is my understanding correct?

Reply:
Yes! So in a the case with only 1 social guardian and only the wallet owner the single social guardian could start recovery, but the wallet owner can cancel it.